### ISSUE TITLE
The function 'homogenize' is not available

### ISSUE THREAD

**Issue by ctjacobs** [firedrakeproject/firedrake#443]
In DOLFIN, I'm doing `homogenize(bc)` (where `bc` is a `DirichletBC` object) which returns a new, homogeneous version of `bc`. This function isn't available in Firedrake.

Patch:

```
diff --git a/firedrake/bcs.py b/firedrake/bcs.py
index 078da48..6c99813 100644
--- a/firedrake/bcs.py
+++ b/firedrake/bcs.py
@@ -222,3 +222,17 @@ class DirichletBC(object):
         self.apply(r)
         if not currently_zeroed:
             self.restore()
+
+def homogenize(bc):
+   """Create a homogeneous version of a :class:`.DirichletBC` object and return it. If
+   ``bc`` is a list or tuple containing one or more :class:`.DirichletBC` objects,
+   then return a list of the homogeneous versions of those :class:`.DirichletBC`s.
+
+   :arg bc: a :class:`.DirichletBC`, or list or tuple comprising :class:`.DirichletBC`(s).
+   """
+   if isinstance(bc, (list, tuple)):
+      bcs = bc
+      return [homogenize(bc) for bc in bcs]
+   V = bc.function_space()
+   homogenized_bc = DirichletBC(V, 0, bc.sub_domain)
+   return homogenized_bc
```

**Comment by dham** [comment]
Fair enough. Please put in a pull request for this change.

**Comment by funsim** [comment]
Nice! This is also crucial for firedrake-adjoint to work correctly with
Dirichlet BCs (because the adjoint BC are the homogenzied' version of the
forward BC).

2015-02-03 22:35 GMT+01:00 David Ham <notifications@github.com>:

> Fair enough. Please put in a pull request for this change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firedrakeproject/firedrake/issues/443#issuecomment-72739837>
> .
>



-- 
Simon Funke

Center for Biomedical Computing
Simula Research Laboratory
Oslo, Norway
simon@simula.no

**Comment by dham** [comment]
Yes, Christian is working on Firedrake adjoint. Apparently it's working
surprisingly easily.

On Tuesday, February 3, 2015, Simon Funke <notifications@github.com> wrote:

>  Nice! This is also crucial for firedrake-adjoint to work correctly with
> Dirichlet BCs (because the adjoint BC are the homogenzied' version of the
> forward BC).
>
> 2015-02-03 22:35 GMT+01:00 David Ham <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>:
>
> > Fair enough. Please put in a pull request for this change.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/firedrakeproject/firedrake/issues/443#issuecomment-72739837>
>
> > .
> >
>
>
>
> --
> Simon Funke
>
> Center for Biomedical Computing
> Simula Research Laboratory
> Oslo, Norway
> simon@simula.no <javascript:_e(%7B%7D,'cvml','simon@simula.no');>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/firedrakeproject/firedrake/issues/443#issuecomment-72748987>
> .
>


-- 
Dr David Ham
Departments of Mathematics and Computing
Imperial College London

http://www.imperial.ac.uk/people/david.ham

**Comment by ctjacobs** [comment]
Indeed. @funsim: I am pushing some homogenize-related fixes to my firedrake-adjoint fork right now. I'll make a pull request for that soon.

**Comment by funsim** [comment]
Exciting!

**Comment by ctjacobs** [comment]
Fixed by #444. Closing.
